src/python/pants/option/options_fingerprinter.py

```diff
@@ -10,16 +10,28 @@
 from builtins import object, open, str
 from hashlib import sha1
 
+from future.moves import collections
+
 from pants.base.build_environment import get_buildroot
 from pants.base.hash_utils import stable_json_hash
 from pants.option.custom_types import (UnsetBool, dict_with_files_option, dir_option, file_option,
                                        target_option)
 
 
+# TODO: this class's methods could be wrapped in @memoized_method -- we would have to manage the
+# cache size somehow if this is done or it might blow up.
 class Encoder(json.JSONEncoder):
+  def _encode_dict_wrapping_keys(self, d):
+    return {self.encode(k):v for k, v in d.items()}
+
   def default(self, o):
     if o is UnsetBool:
       return '_UNSET_BOOL_ENCODING'
+    elif isinstance(o, collections.Iterable):
+      if isinstance(o, collections.Mapping):
+        return self._encode_dict_wrapping_keys(o)
```

Comment:
Probably makes sense to just inline `_encode_dict_wrapping_keys`.

Reply:
It may also be worth a note that this encoding is 1-way, but that's OK since we never need to decode, just hash the encoding.

Reply:
I ended up doing the string conversion that python docs *say* is already done in `json.dumps` in the `fingerprint` method in cb7dfba, which seems to be the path of least resistance (I also added testing). If it's preferable to have `stable_json_sha1()` be able to handle this case for other uses, I can implement that easily.